packages/dom-adapters/src/caret/utils/useSelectionChange.ts

```diff
@@ -0,0 +1,143 @@
+import type { CaretAdapter } from '../CaretAdapter.js';
+import { createSingleton } from './singleton.js';
+
+/**
+ * Singleton that watches for document "selection change" event and delegates the provided callbacks to subscribers.
+ */
+export interface Subscriber {
+  /**
+   * Callback that will be called on "selection change" event.
+   *
+   * @param selection - current document selection
+   */
+  callback: (selection: Selection | null) => void;
+
+  /**
+   * Used to save context of the callback.
+   */
+  context: CaretAdapter;
```

Comment:
It might be not only CaretAdapter in the future I suppose. Could be just unknown